src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs

```diff
@@ -409,165 +409,217 @@ public void RequestLicKey(int dwReserved, [MarshalAs(UnmanagedType.BStr)] out st
+        // If we are being entered because of a call to ICF::CreateInstance(),
+        // "isDesignTime" will be "true".
+        //
+        // If we are being entered because of a call to ICF::CreateInstanceLic(),
+        // "isDesignTime" will be "false" and "key" will point to a non-null
+        // license key.
+        public object AllocateAndValidateLicense(Type type, string key, bool isDesignTime)
+        {
+            object[] parameters;
+            object licContext;
+            if (isDesignTime)
+            {
+                parameters = new object[] { type };
+                licContext = _createDesignContext.Invoke(null, BindingFlags.DoNotWrapExceptions, binder: null, parameters: parameters, culture: null);
             }
-
-            // The CLR invokes this whenever a COM client invokes
-            // IClassFactory2::RequestLicKey on a managed class.
-            public string RequestLicKey(Type type)
+            else
             {
-                // License will be null, since we passed no instance,
-                // however we can still retrieve the "first" license
-                // key from the file. This really will only
-                // work for simple COM-compatible license providers
-                // like LicFileLicenseProvider that don't require the
-                // instance to grant a key.
+                parameters = new object[] { type, key };
+                licContext = _createRuntimeContext.Invoke(null, BindingFlags.DoNotWrapExceptions, binder: null, parameters: parameters, culture: null);
+            }
 
-                // Types are as follows:
-                // LicenseContext, Type, out License, out string
-                var parameters = new object[] { /* use global LicenseContext */ null, type, /* out */ null, /* out */ null };
-                bool isValid = (bool)_validateTypeAndReturnDetails.Invoke(null, BindingFlags.DoNotWrapExceptions, binder: null, parameters: parameters, culture: null);
-                if (!isValid)
-                {
-                    throw new COMException(); //E_FAIL
-                }
+            try
+            {
+                parameters = new object[] { type, licContext };
+                return _createWithContext.Invoke(null, BindingFlags.DoNotWrapExceptions, binder: null, parameters: parameters, culture: null);
+            }
+            catch (Exception exception) when (exception.GetType() == s_licenseExceptionType)
+            {
+                const int CLASS_E_NOTLICENSED = unchecked((int)0x80040112);
+                throw new COMException(exception.Message, CLASS_E_NOTLICENSED);
+            }
+        }
 
-                var license = (IDisposable)parameters[2];
-                if (license != null)
-                {
-                    license.Dispose();
-                }
+        // See usage in native RCW code
+        public void GetCurrentContextInfo(RuntimeTypeHandle rth, ref int fDesignTime, ref IntPtr bstrKey)
```

Comment:
Why not pass `fDesignTime` as `out bool fDesignTime` ?

Reply:
Sure.